File: Xjie/Xjie/Models/ChatModels.swift
import Foundation

// MARK: - 文献引用

/// 与后端 schemas/literature.py CitationBundle 对应
struct Citation: Codable, Identifiable, Hashable, Equatable {
    let claim_id: Int
    let literature_id: Int
    let claim_text: String
    let evidence_level: String      // L1 / L2 / L3 / L4
    let short_ref: String           // e.g. "Segal et al., Cell 2015"
    let journal: String?
    let year: Int?
    let sample_size: Int?
    let confidence: String          // high / medium / low
    let score: Double?

    var id: Int { claim_id }
}

// MARK: - 聊天

struct ChatConversation: Codable, Identifiable {
    let id: String
    let title: String?
    let message_count: Int?
    let updated_at: String?
    let created_at: String?
}

/// BUG-01 FIX: id 改为存储属性，避免 computed property 每次访问生成新 UUID
/// 导致 SwiftUI ForEach 无限刷新
struct ChatMessage: Decodable, Identifiable {
    let id: String
    let role: String
    let content: String
    let analysis: String?
    let created_at: String?
    let citations: [Citation]

    enum CodingKeys: String, CodingKey {
        case id, role, content, analysis, created_at, citations
    }

    init(id: String = UUID().uuidString, role: String, content: String, analysis: String? = nil, created_at: String? = nil, citations: [Citation] = []) {
        self.id = id
        self.role = role
        self.content = content
        self.analysis = analysis
        self.created_at = created_at
        self.citations = citations
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        self.id = (try? container.decode(String.self, forKey: .id)) ?? UUID().uuidString
        self.role = try container.decode(String.self, forKey: .role)
        self.content = try container.decode(String.self, forKey: .content)
        self.analysis = try? container.decode(String.self, forKey: .analysis)
        self.created_at = try? container.decode(String.self, forKey: .created_at)
        self.citations = (try? container.decode([Citation].self, forKey: .citations)) ?? []
    }
}

struct ChatRequest: Encodable {
    let message: String
    let thread_id: String?
}

struct ChatResponse: Codable {
    let summary: String?
    let analysis: String?
    let answer_markdown: String?
    let confidence: Double?
    let followups: [String]?
    let thread_id: String?
    let citations: [Citation]?

    init(summary: String? = nil,
         analysis: String? = nil,
         answer_markdown: String? = nil,
         confidence: Double? = nil,
         followups: [String]? = nil,
         thread_id: String? = nil,
         citations: [Citation]? = nil) {
        self.summary = summary
        self.analysis = analysis
        self.answer_markdown = answer_markdown
        self.confidence = confidence
        self.followups = followups
        self.thread_id = thread_id
        self.citations = citations
    }
}

// MARK: - 授权

struct ConsentUpdate: Encodable {
    let allow_ai_chat: Bool
}

struct ConsentResponse: Decodable {
    let allow_ai_chat: Bool
    let allow_data_upload: Bool?
}

File: Android/app/src/main/java/com/xjie/app/core/model/ChatModels.kt
package com.xjie.app.core.model

import kotlinx.serialization.Serializable

@Serializable
data class Citation(
    val claim_id: Int,
    val literature_id: Int,
    val claim_text: String,
    val evidence_level: String,
    val short_ref: String,
    val journal: String? = null,
    val year: Int? = null,
    val sample_size: Int? = null,
    val population: String? = null,
    val study_design: String? = null,
    val confidence: String,
    val score: Double? = null,
)

@Serializable
data class ChatConversation(
    val id: String,
    val title: String? = null,
    val message_count: Int? = null,
    val updated_at: String? = null,
    val created_at: String? = null,
)

@Serializable
data class ChatMessage(
    val id: String,
    val role: String,
    val content: String,
    val analysis: String? = null,
    val created_at: String? = null,
    val citations: List<Citation> = emptyList(),
)

@Serializable
data class ChatRequest(
    val message: String,
    val thread_id: String? = null,
    val client_message_id: String? = null,
)

@Serializable
data class ChatInteractionRoute(
    val version: String,
    val route_id: String,
    val strategy: String,
    val primary_intent: String,
    val depth: String,
    val safety_level: String,
    val subject_type: String,
    val needs_literature: Boolean,
    val max_followups: Int,
    val progress_steps: List<String> = emptyList(),
)

@Serializable
data class ChatResponse(
    val summary: String? = null,
    val analysis: String? = null,
    val answer_markdown: String? = null,
    val confidence: Double? = null,
    val followups: List<String>? = null,
    val thread_id: String? = null,
    val message_id: String? = null,
    val response_state: String? = null,
    val interaction_route: ChatInteractionRoute? = null,
    val quality_flags: List<String>? = null,
    val citations: List<Citation>? = null,
)

@Serializable
data class ChatStreamEnvelope(
    val type: String,
    val route: ChatInteractionRoute? = null,
    val step: String? = null,
    val delta: String? = null,
    val result: ChatResponse? = null,
    val message: String? = null,
    val retryable: Boolean? = null,
)

sealed interface ChatStreamEvent {
    data class Route(val route: ChatInteractionRoute) : ChatStreamEvent
    data class Progress(val step: String) : ChatStreamEvent
    data class Token(val delta: String) : ChatStreamEvent
    data class Done(val result: ChatResponse) : ChatStreamEvent
}

@Serializable
data class ConsentUpdate(val allow_ai_chat: Boolean)

@Serializable
data class ConsentResponse(
    val allow_ai_chat: Boolean,
    val allow_data_upload: Boolean? = null,
)
